[#4] 회원 가입 기능 구현

tilsong opened the pull request:
## PR 유형

- [ ] 버그 수정
- [x] 기능
- [ ] 코드 스타일
- [ ] 리팩토링
- [ ] 빌드
- [ ] CI/CD
- [ ] 문서
- [ ] 기타
  <br>

## 변경사항

### 변경 전
- 기본 설정 코드만 존재.
### 변경 후
- validation과 트랜잭션을 이용한 테스트를 위해 관련 의존성이 추가되었습니다.
- 회원 가입 기능이 추가되었습니다.
- 비밀번호 암호화 기능이 추가되었습니다.

## 기타
-  아래의 의존성들이 추가되었습니다. 커밋 내용을 통해 의존성 추가 이유를 확인하실 수 있습니다.
    - spring-boot-starter-validation
    - org.springframework.boot:spring-boot-starter-data-jpa
    - com.h2database:h2
    - group: 'org.mindrot', name: 'jbcrypt', version: '0.4'

tilsong committed:
[#4] 회원 가입 기능 구현 - Controller

- 요청 데이터에 대한 Validation 적용을 위한 "spring-boot-starter-validation" 의존성 추가
- 회원 가입 dto 생성
  - erd 데이터 참고
  - validation 적용
- 요청 데이터의 validation 예외를 처리하기 위한 전역 예외 핸들러(GlobalExceptionHandler) 및 handleMethodArgumentNotValid 메서드 추가
- Controller에 회원 가입 요청을 받아 응답해주는 signUp 메서드 추가

tilsong committed:
[#4] 회원 가입 기능 구현 - Controller

- Controller에 아이디 중복 체크 요청을 받아 응답하는 isDuplicateId 메서드 추가

tilsong committed:
[#4] 회원 가입 기능 구현 - Service, Repository

- @Transactional 적용을 위한 "org.springframework.boot:spring-boot-starter-data-jpa", "com.h2database:h2" 의존성 추가
- 아이디 중복 체크 시 발생하는 DuplicateIdException 예외 클래스 추가 및 전역 예외 핸들러에 해당 예외를 메서드로 적용
- 회원 가입 서비스를 제공하는 SignUpService 클래스 생성 - 회원 가입 및 아이디 중복 체크 메서드 추가
- 유저 도메인 추가 (jpa 관련 어노테이션은 추가되지 않음)
- 도메인 객체 저장을 위한 Repository 인터페이스와 구현 클래스 생성. 구현 클래스는 Adapter 패턴으로 구현됨.
  - 테스트를 위한 Map을 사용한 더미 데이터 추가
- 요청 데이터의 validation 예외를 처리하기 위한 전역 예외 핸들러(GlobalExceptionHandler) 및 handleMethodArgumentNotValid 메서드 추가
- Controller에 회원 가입 요청을 받아 응답해주는 signUp 메서드 추가

puallab reviewed (approved):
좋습니다!

tilsong committed:
[#4] 회원 가입 기능 구현 - Service, Password 암호화

- 비밀번호 암호화를 위한 ` group: 'org.mindrot', name: 'jbcrypt', version: '0.4' ` 의존성 추가
- 비밀번호 관리를 위한 PasswordManager 인터페이스 및 구현체 추가
  - jbcrypt 라이브러리 사용
  - SignUpService에서 유저 객체를 생성할 때 PasswordManager의 비밀번호 암호화 메서드를 사용해 생성한 값을 저장

puallab commented on src/main/java/com/directors/domain/user/User.java:
접근제어자를 package private으로 하신 이유가 있을까요? builder와 관련 있는건가요?

f-lab-fi replied:
인터페이스 메서드의 디폴트 가시성은 package private이 아니라 public 아닌가요? ㅎㅎ

f-lab-fi commented on src/main/java/com/directors/infrastructure/exception/user/DuplicateIdException.java:
```diff
@@ -0,0 +1,7 @@
+package com.directors.infrastructure.exception.user;
+
+public class DuplicateIdException extends RuntimeException {
+    public DuplicateIdException(String message) {
+        super(message);
+    }
+}
```
예외를 정의할 때는 **어떤 값이 예외를 유발했는지 함께 저장**하는 쪽이 응집력 있는 코드를 만들 여지가 많아집니다.
예외는 그 자체로 예외를 유발한 값과 함께 던져지고, 예외를 핸들링하는 쪽에서는 로그를 남기는 형태가 유리합니다.

은석님이 지금 이 코드에서 어떤 id가 들어왔길래 중복이 발생한건지 로그를 남긴다면 어디에 남길지 고민해보시면 바로 문제가 드러날겁니다.

f-lab-fi commented on src/main/java/com/directors/application/user/SignUpService.java:
```diff
@@ -0,0 +1,41 @@
+        userRepository.saveUser(newUser);
+    }
+
+    @Transactional
+    public void isDuplicatedUser(String id) {
+        if (userRepository.findUserById(id) != null) {
+            throw new DuplicateIdException("이미 존재하는 Id입니다");
+        }
+    }
```
DuplicateIdException 라는 예외가 "이미 존재하는 Id입니다" 라는 메시지 말고 다른 메시지가 들어올 여지가 있을까요?
그럼 예외를 던져주기 위한 DuplicateIdException 생성자마다 불필요하게 동일한 메시지를 넣어주고 있는거 아닌지 고민해봐야 할 것 같습니다.

f-lab-fi commented on src/main/java/com/directors/application/user/SignUpService.java:
```diff
@@ -0,0 +1,41 @@
+    public void signUp(SignUpRequest signUpRequest) {
+        isDuplicatedUser(signUpRequest.userId());
+
+        User newUser = User.builder()
+                .userId(signUpRequest.userId())
+                .password(pm.encodePassword(signUpRequest.password()))
+                .name(signUpRequest.name())
+                .nickname(signUpRequest.nickname())
+                .email(signUpRequest.email())
+                .phoneNumber(signUpRequest.phoneNumber())
+                .build();
```
DTO <-> Entity 간 변환은 대부분의 웹 애플리케이션에서 공통적으로 나타나는 코드입니다.
물론 지금 은석님 코드에서는 Record를 사용하셨는데요, Record가 DTO의 역할을 하고 있는거죠.
그런데 이렇게 DTO <-> Entity 간 변환을 하는 과정에서 매번 이렇게 Builder를 통해 변환해줘야할까요?
이러면 빌더 코드가 너무 여러곳에 중복되어 발생하는게 아닐까요?
그렇게 중요한 코드가 아닌데, 계속 반복되면 코드의 가독성을 떨어뜨릴 것 같습니다.

f-lab-fi commented on src/main/java/com/directors/infrastructure/jpa/user/UserRepositoryAdapter.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@Repository
+public class UserRepositoryAdapter implements UserRepository {
```
얘는 왜 구현체 이름이 어댑터인가요? ㅎㅎ

f-lab-fi commented on src/main/java/com/directors/presentation/user/UserController.java:
```diff
@@ -1,14 +1,36 @@
+        return new ResponseEntity(HttpStatus.CREATED);
     }
 
+    @PostMapping("/duplicated/{id}")
+    public ResponseEntity<HttpStatus> isDuplicateId(
+            @PathVariable @NotBlank(message = "입력 값이 존재하지않습니다.")
+            @Size(min = 8, max = 20, message = "아이디의 길이가 8-20글자 사이로 입력되지 않았습니다.") String id
+    ) {
```
이 API는 POST 메서드일 이유가 있을까요?
어떤 의도가 있는지 궁금합니다.

tilsong committed:
[#4] 회원 가입 기능 구현 - 피드백 반영: UserRepository 구현 클래스명 변경

- UserRepository의 구현 클래스명을 보다 명확한 이름으로 변경했습니다.
  - UserRepositoryAdapter -> InmemoryUserRepository

tilsong committed:
[#4] 회원 가입 기능 구현 - 피드백 반영: UserController

- 메서드 isDuplicateId는 조회의 의미를 가지므로 @Post -> @Get으로 변경했습니다.

tilsong committed:
[#4] 회원 가입 기능 구현 - 피드백 반영:  Dto -> Entity 간 변환, @Builder 패턴 제거

- User 도메인 객체 생성 시 생성자를 사용하고, Builder 관련 코드를 제거했습니다.
- Dto -> Entity 변환을 위해 UserConverter 클래스를 추가했습니다.

tilsong committed:
[#4] 회원 가입 기능 구현 - 피드백 반영:  중복 예외 관련

- 특정 예외 발생 시 사용될 메시지가 고정되어 있으므로, 이를 상수로 예외 클래스 내부에 추가했습니다.
- 예외 발생의 원인을 예외 객체에 담고, 이를 핸들러에서 로그로 추가하도록 했습니다.

f-lab-fi commented on src/main/java/com/directors/application/user/UserConverter.java:
```diff
@@ -0,0 +1,18 @@
+import com.directors.domain.user.User;
+import com.directors.presentation.user.request.SignUpRequest;
+
+public class UserConverter {
+    public static User toEntity(SignUpRequest signUpRequest) {
+        // region은 회원 가입 후, 추가로 설정할 수 있는 로직을 만들 예정입니다.
+        return new User(
+                signUpRequest.userId(),
+                signUpRequest.password(),
+                signUpRequest.name(),
+                signUpRequest.nickname(),
+                signUpRequest.email(),
+                signUpRequest.phoneNumber(),
+                "", 0);
+    }
+}
```
Converter를 추가하는 쪽으로 바꾸셨군요.
다른 방법도 있긴 합니다. 바로 DTO 쪽에 User를 반환하는 정적 팩토리 메서드를 추가하는 방법도 가능합니다.

상황에 따라 이렇게 Converter를 따로 만들지, DTO(SignUpRequest) 쪽에 정적 팩토리 메서드를 만들지 고민해보시기 바랍니다. :)

f-lab-fi reviewed (approved):
수고하셨습니다.

tilsong committed:
[#4] 회원 가입 기능 구현 - 비밀번호 암호화 코드 수정

- 회원가입을 위한 유저 객체 생성 후, 패스워드를 암호화 해주는 로직을 추가했습니다.
- 패스워드 암호화 메서드의 이름이 보다 명확한 의미를 담도록 이을 수정했습니다.
